pkg/codegen/schema/pulumi.json

```diff
@@ -0,0 +1,509 @@
+                "$ref": "#/$defs/complexTypeSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "provider": {
+            "description": "The provider type for this package.",
+            "$ref": "#/$defs/resourceSpec"
+        },
+        "resources": {
+            "description": "A map from type token to resourceSpec that describes the set of resources and components defined by this package.",
+            "type": "object",
+            "additionalProperties": {
+                "$ref": "#/$defs/resourceSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "functions": {
+            "description": "A map from token to functionSpec that describes the set of functions defined by this package.",
+            "type": "object",
+            "additionalProperties": {
+                "$ref": "#/$defs/functionSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "language": {
+            "description": "Additional language-specific data about the package.",
+            "type": "object"
+        }
+    },
+    "additionalProperties": false,
+    "required": [
+        "name"
+    ],
+    "$defs": {
+        "token": {
+            "title": "Token",
+            "type": "string",
+            "$comment": "In the regex below, the 'module' portion of the token is optional. However, a missing module component creates a '::', which breaks URNs ('::' is the URN delimiter). We have many test schemas that use an empty module component successfully, as they never create URNs; while these are _probably_ the only places that need updating, it might be possible that there are module-less type tokens in the wild elsewhere and we may need to remain compatible with those tokens.",
+            "pattern": "^[^0-9][-a-zA-Z0-9]*:([^0-9][a-zA-Z0-9._/]*)?:[^0-9][a-zA-Z0-9._/]*$"
+        },
+        "typeSpec": {
+            "title": "Type Reference",
+            "description": "A reference to a type",
+            "type": "object",
+            "properties": {
+                "plain": {
+                    "description": "Indicates that when used as an input, this type does not accept eventual values.",
+                    "type": "boolean"
+                }
+            },
+            "oneOf": [
+                {
+                    "title": "Primitive Type",
+                    "description": "A reference to a primitive type.",
+                    "type": "object",
+                    "properties": {
+                        "type": {
+                            "description": "The primitive type, if any",
+                            "type": "string",
+                            "enum": ["boolean", "integer", "number", "string"]
+                        },
+                        "additionalProperties": false,
```

Comment:
I had trouble parsing this. What is this for? Note it's in the "properties" context. Should it be moved higher up to the object x where x.description == "A reference to a primitive type"?

Reply:
Ah, it should be up one level, yes. Good catch!

Reply:
If it's actually "additionalProperties" in the object language, what does false mean in the meta-language?

Reply:
Oh, wait, I take that back

Reply:
This line prevents the schema from matching if the object has a property named `additionalProperties`

Reply:
This gets a bit hard to read because the names used by the metaschema to describe the package schema and the names used by the package schema to describe a package overlap.

Reply:
Okay. So 'false' in the JSON schema meta-language means "never match". This makes some sense. I'll need to adjust the coverage script to take care of this.

Reply:
Yep. I've updated the description of the type reference schema to explain how types are discriminated.